File: server/src/modules/cart/cart.mapper.ts
import { resolvePublicProductImagePath } from "../images/image.service";
import type { Product } from "../products/product.types";
import type { CartDto, CartItemDto, CartItemRecord, CartItemRow } from "./cart.types";

export interface CartMapperOptions {
  publicImageBaseUrl?: string;
}

export function mapCartItemRowToRecord(row: CartItemRow): CartItemRecord {
  return {
    id: row.id,
    userKey: row.user_key,
    productId: row.product_id,
    quantity: row.quantity,
    selected: row.selected,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function mapCartToDto(
  items: CartItemRecord[],
  products: Product[],
  options: CartMapperOptions = {},
): CartDto {
  const productsById = new Map(
    products.map((product) => [product.id, product]),
  );
  const dtoItems = items.flatMap((item): CartItemDto[] => {
    const product = productsById.get(item.productId);

    if (!product) {
      return [];
    }

    const priceCents = product.basePriceCents;
    const subtotalCents = priceCents * item.quantity;

    return [{
      id: item.id,
      productId: product.id,
      name: product.name,
      brand: product.brand,
      category: product.category,
      priceCents,
      priceText: formatPriceText(priceCents, product.currency),
      quantity: item.quantity,
      selected: item.selected,
      subtotalCents,
      available: isProductAvailable(product),
      tags: product.visualTags,
      imagePath: resolvePublicProductImagePath(
        product.imagePath,
        options.publicImageBaseUrl,
      ),
    }];
  });

  return {
    items: dtoItems,
    summary: {
      totalCount: dtoItems.reduce((sum, item) => sum + item.quantity, 0),
      selectedCount: dtoItems
        .filter((item) => item.selected)
        .reduce((sum, item) => sum + item.quantity, 0),
      selectedTotalCents: dtoItems
        .filter((item) => item.selected)
        .reduce((sum, item) => sum + item.subtotalCents, 0),
      currency: "CNY",
    },
  };
}

export function isProductAvailable(product: Product): boolean {
  if (product.status !== "active") {
    return false;
  }

  return product.skus.length === 0 || product.skus.some((sku) => sku.available);
}

function formatPriceText(priceCents: number, currency: string): string {
  const amount = priceCents % 100 === 0
    ? `${priceCents / 100}`
    : (priceCents / 100).toFixed(2);

  if (currency.toUpperCase() === "CNY") {
    return `¥${amount}`;
  }

  return `${currency.toUpperCase()} ${amount}`;
}

File: client/android/app/src/main/java/com/shopmate/app/data/cart/CartMapper.kt
package com.shopmate.app.data.cart

import com.shopmate.app.data.network.ShopMateImageUrlResolver
import com.shopmate.app.data.products.formatCnyCentsText
import com.shopmate.app.data.products.resolveProductPlaceholder
import com.shopmate.app.ui.cart.CartContentUi
import com.shopmate.app.ui.cart.CartSummaryUi
import com.shopmate.app.ui.model.CartItemUi
import com.shopmate.app.ui.model.ProductCardUi

fun CartDto.toCartContentUi(
    imageUrlResolver: ShopMateImageUrlResolver? = null,
): CartContentUi =
    CartContentUi(
        items = items.map { item -> item.toCartItemUi(imageUrlResolver) },
        summary = summary.toCartSummaryUi(),
    )

private fun CartItemDto.toCartItemUi(
    imageUrlResolver: ShopMateImageUrlResolver?,
): CartItemUi =
    CartItemUi(
        id = id,
        product = ProductCardUi(
            id = productId,
            name = name.ifBlank { "未命名商品" },
            priceText = priceText.ifBlank { formatCnyCentsText(priceCents) },
            imageRes = resolveProductImageRes(),
            tags = tags.filter { tag -> tag.isNotBlank() }.take(MAX_CART_TAGS),
            recommendationReason = buildCartReason(),
            imageUrl = imageUrlResolver?.resolve(imagePath),
        ),
        quantity = quantity.coerceAtLeast(1),
        subtotalText = formatCnyCentsText(subtotalCents),
        selected = selected,
        available = available,
    )

private fun CartSummaryDto.toCartSummaryUi(): CartSummaryUi =
    CartSummaryUi(
        totalCount = totalCount,
        selectedCount = selectedCount,
        selectedTotalCents = selectedTotalCents,
        selectedTotalText = formatCnyCentsText(selectedTotalCents),
        currency = currency,
    )

private fun CartItemDto.buildCartReason(): String {
    val brandText = brand.orEmpty().trim()
    val categoryText = category.orEmpty().trim()
    return listOf(brandText, categoryText)
        .filter { value -> value.isNotBlank() }
        .joinToString(" · ")
        .ifBlank { "已加入购物车，可调整数量和选择状态。" }
}

private fun CartItemDto.resolveProductImageRes(): Int {
    return resolveProductPlaceholder(
        listOf(productId, name, brand, category, tags.joinToString(" "), imagePath),
    )
}

private const val MAX_CART_TAGS = 2
